File: src/main/java/com/github/iamniklas/liocore/procedures/variants/RainbowMonoProcedure.java
package com.github.iamniklas.liocore.procedures.variants;

import com.github.iamniklas.colorspaces.ColorHSV;
import com.github.iamniklas.liocore.led.LEDDataBundle;
import com.github.iamniklas.liocore.led.colorspace.LIOColor;
import com.github.iamniklas.liocore.network.LEDUpdateModel;
import com.github.iamniklas.liocore.procedures.Procedure;

public class RainbowMonoProcedure extends Procedure {

    private ColorHSV colorHsv = new ColorHSV(0, 1.0f, 1.0f);

    public RainbowMonoProcedure(LEDUpdateModel _ledUpdateModel) {
        super(_ledUpdateModel);
    }

    @Override
    public void start() {
        ledUpdateModel.bundle.speed = Math.max(1.0f, ledUpdateModel.bundle.speed);
    }

    @Override
    public void update() {
        colorHsv.h = colorHsv.h > 360.0f ? 0 : (int) (colorHsv.h + ledUpdateModel.bundle.speed);
        strip.setAllPixels(LIOColor.fromHSV(colorHsv));
    }

    @Override
    public void updateLEDDataBundle(LEDDataBundle ledDataBundle) {

    }

    @Override
    public boolean validateBundleData() {
        return ledUpdateModel.bundle.speed != null;
    }
}

File: src/main/java/com/github/iamniklas/liocorekotlin/procedures/RainbowMonoProcedure.kt
package com.github.iamniklas.liocorekotlin.procedures

import com.github.iamniklas.liocorekotlin.led.ColorHSV
import com.github.iamniklas.liocorekotlin.led.LEDDataBundle

class RainbowMonoProcedure(_bundle: com.github.iamniklas.liocorekotlin.led.LEDDataBundle) : com.github.iamniklas.liocorekotlin.procedures.Procedure(_bundle) {
    var mColorHSV = com.github.iamniklas.liocorekotlin.led.ColorHSV(0, 1.0f, 1.0f)
    var mHueCounter = 0
    var mRepetitions = 1f
    var mSpeed = 1f
    override fun start() {}
    public override fun update() {
        mColorHSV.h = (if (mColorHSV.h > 360.0f) 0 else mColorHSV.h + mSpeed).toInt()
        mStrip!!.setAllPixels(mColorHSV.ToRGB().toSystemColor())
    }

    init {
        mSpeed = _bundle.speed!!
    }
}